swap/src/database/eigensync.rs

```diff
@@ -0,0 +1,386 @@
+use std::{collections::{HashMap, HashSet}, str::FromStr, sync::Arc, time::{Duration, Instant}};
+
+use autosurgeon::{Hydrate, HydrateError, Reconcile, Reconciler};
+use autosurgeon::reconcile::NoKey;
+use bdk::bitcoin::hashes::hash160::Hash;
+use eigensync::EigensyncHandle;
+use libp2p::{Multiaddr, PeerId};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+// serde kept via Cargo features; no direct derives used here
+use tokio::sync::RwLock;
+use uuid::Uuid;
+use rust_decimal::Decimal;
+use anyhow::anyhow;
+use time::UtcDateTime;
+
+use crate::{database::Swap, monero::{LabeledMoneroAddress, MoneroAddressPool, TransferProof}, protocol::{Database, State}};
+
+#[derive(Serialize, Deserialize, Clone, Eq, Hash, PartialEq, Hydrate, Reconcile, Debug)]
+struct KeyWrapper<T>(T, String);
+
+impl<T: Serialize> KeyWrapper<T> {
+    fn new(key: T) -> Self {
+        let json = serde_json::to_string(&key).unwrap();
+        Self(key, json)
+    }
+}
+
+impl<T> AsRef<str> for KeyWrapper<T> {
+    fn as_ref(&self) -> &str {
+        &self.1
+    }
+}
+
+impl<T: DeserializeOwned> From<String> for KeyWrapper<T> {
+    fn from(s: String) -> Self {
+        Self(serde_json::from_str(&s).unwrap(), s)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct EigensyncDocument {
+    // swap_id, swap
+    states: HashMap<InnerStateKey, State>, // swap_states table
+    // peer_addresses table
+    peer_addresses: HashMap<InnerPeerAddressesKey, ()>, // (peer_id, address)
+    // peers table
+    peers: HashMap<Uuid, PeerId>, //  (swap_id, peer_id)
+    // monero_addresses table
+    monero_addresses: HashMap<InnerMoneroAddressKey, MoneroAddressValue>, // (swap_id, address) -> (percentage, label)
+    // buffered_transfer_proofs table
+    buffered_transfer_proofs: HashMap<Uuid, TransferProof>, // (swap_id, proof)
+}
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
+struct MoneroAddressValue(Decimal, String);
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
+struct InnerMoneroAddressKey(Uuid, Option<String>);
+
+type MoneroAddressKey = KeyWrapper<InnerMoneroAddressKey>;
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
+struct InnerStateKey(Uuid, UtcDateTime);
+
+type StateKey = KeyWrapper<InnerStateKey>;
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
+struct InnerPeerAddressesKey(PeerId, Multiaddr);
+
+type PeerAddressesKey = KeyWrapper<InnerPeerAddressesKey>;
+
+#[derive(Debug, Clone, Reconcile, Hydrate, PartialEq, Default)]
+struct EigensyncWire {
+    states: HashMap<StateKey, String>,
+    // encode (peer_id, addr) -> "peer_id|addr", unit value as bool true
+    peer_addresses: HashMap<PeerAddressesKey, bool>,
+    // swap_id -> peer_id
+    peers: HashMap<String, String>,
+    // encode (swap_id, address?) -> "swap_id|address_or_-"; store (Decimal, String) as (String, String)
```

Comment:
these comments are not up to date anymore?

We are not serialiazing as string anymore